src/event_pusher/mod_event_pusher_rabbit.erl

```diff
@@ -0,0 +1,231 @@
+%%% Definitions
+%%%===================================================================
+
+-define(DEFAULT_PRESENCE_EXCHANGE, <<"presence">>).
+-define(DEFAULT_PRESENCE_EXCHANGE_TYPE, <<"topic">>).
+-define(DEFAULT_CHAT_MSG_EXCHANGE, <<"chat_msg">>).
+-define(DEFAULT_CHAT_MSG_EXCHANGE_TYPE, <<"topic">>).
+-define(DEFAULT_CHAT_MSG_SENT_TOPIC, <<"chat_msg_sent">>).
+-define(DEFAULT_CHAT_MSG_RECV_TOPIC, <<"chat_msg_recv">>).
+-define(DEFAULT_GROUP_CHAT_MSG_EXCHANGE, <<"groupchat_msg">>).
+-define(DEFAULT_GROUP_CHAT_MSG_EXCHANGE_TYPE, <<"topic">>).
+-define(DEFAULT_GROUP_CHAT_MSG_SENT_TOPIC, <<"groupchat_msg_sent">>).
+-define(DEFAULT_GROUP_CHAT_MSG_RECV_TOPIC, <<"groupchat_msg_recv">>).
+
+%%%===================================================================
+%%% Exports
+%%%===================================================================
+
+%% MIM module callbacks
+-export([start/2, stop/1]).
+
+%% API
+-export([push_event/3]).
+
+%%%===================================================================
+%%% Callbacks
+%%%===================================================================
+
+-spec start(Host :: jid:server(), Opts :: proplists:proplist()) -> ok.
+start(Host, _Opts) ->
+    initialize_metrics(Host),
+    create_exchanges(Host),
+    ok.
+
+-spec stop(Host :: jid:server()) -> ok.
+stop(_Host) ->
+    ok.
+
+push_event(Acc, _, #user_status_event{jid = UserJID, status = Status}) ->
+    handle_user_presence_change(UserJID, Status),
+    Acc;
+push_event(Acc, _, ChatEvent = #chat_event{}) ->
+    handle_user_chat_event(ChatEvent),
+    Acc;
+push_event(Acc, _, _) ->
+    Acc.
+
+%%%===================================================================
+%%% Internal functions
+%%%===================================================================
+
+-spec initialize_metrics(Host :: jid:server()) -> [ok | {ok | error, term()}].
+initialize_metrics(Host) ->
+    [mongoose_metrics:ensure_metric(Host, Name, Type)
+     || {Name, Type} <- mongoose_rabbit_worker:list_metrics()].
+
+-spec create_exchanges(Host :: jid:server()) -> ok.
+create_exchanges(Host) ->
+    Res =
+        [call_rabbit_worker(Host, {amqp_call,
+                                   mongoose_amqp:exchange_declare(ExName, Type)})
+         || {ExName, Type} <- exchanges(Host)],
+    verify_exchanges_were_created_or_crash(Res).
+
+-spec handle_user_presence_change(JID :: jid:jid(), Status :: atom()) -> ok.
+handle_user_presence_change(JID = #jid{lserver = Host}, Status) ->
+    Exchange = exchange_opt(Host, presence_exchange, name,
+                            ?DEFAULT_PRESENCE_EXCHANGE),
+    RoutingKey = presence_routing_key(JID),
+    Message = presence_msg(JID, Status),
+    PublishMethod = mongoose_amqp:basic_publish(Exchange, RoutingKey),
+    AMQPMessage = mongoose_amqp:message(Message),
+    cast_rabbit_worker(Host, {amqp_publish, PublishMethod, AMQPMessage}).
+
+-spec handle_user_chat_event(#chat_event{}) -> ok.
+handle_user_chat_event(#chat_event{type = normal}) -> ok;
```

Comment:
It's better to match for only `chat` and `groupchat` explicitly and ignore everything else. Right now, this function will crash if after some changes in the future, a `headline` messages reaches this function.